tracer/src/Datadog.Trace/AppSec/BodyExtractor.cs

```diff
@@ -0,0 +1,180 @@
+// <copyright file="BodyExtractor.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text.RegularExpressions;
+using Datadog.Trace.AppSec.Waf;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.AppSec
+{
+    internal static class BodyExtractor
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(BodyExtractor));
+
+        private static readonly Regex NameExtractor = new Regex(@"\<(?'PropertyName'[^\>]+)\>", RegexOptions.Compiled);
+
+        private static readonly HashSet<Type> AdditionalPrimitives = new()
+        {
+            typeof(string),
+            typeof(decimal),
+            typeof(Guid),
+            typeof(DateTime),
+            typeof(DateTimeOffset),
+            typeof(TimeSpan)
+        };
+
+        internal static object GetKeysAndValues(object body)
```

Comment:
You return either a collection or a string, I assume you could use a stronger type.
It could be useful to preallocate the event dictionary in InstrumentationGateway for instance.

nit-if the body is a string you don't return KeysAndValues. I'd rename to `ParseBody` or `ExtractPropertiesFromBody`?

Reply:
I renamed the method to `Extract`.

Don't see how you can type something as a string or a dictionary effectively without union types. It not really useful here, since the resulting object is added directly to a dictionary of type `Dictionary<string, object>`.

I used pre-allocation of the dictionaries in dd3e5c554, but it doesn't seem to change much:

```
Before

|                       Method |        Job |       Runtime |     Toolchain |     Mean |    Error |   StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------------- |----------- |-------------- |-------------- |---------:|---------:|---------:|------:|--------:|-------:|------:|------:|----------:|
| BodyExtractorMoreComplexBody | Job-PHWFYO |    .NET 4.7.2 |        net472 |       NA |       NA |       NA |     ? |       ? |      - |     - |     - |         - |
| BodyExtractorMoreComplexBody | Job-PKDTCC | .NET Core 3.1 | netcoreapp3.1 | 16.56 us | 0.282 us | 0.236 us |     ? |       ? | 0.0493 |     - |     - |    7576 B |


After

|                       Method |        Job |       Runtime |     Toolchain |     Mean |    Error |   StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------------- |----------- |-------------- |-------------- |---------:|---------:|---------:|------:|--------:|-------:|------:|------:|----------:|
| BodyExtractorMoreComplexBody | Job-NBCJCQ |    .NET 4.7.2 |        net472 |       NA |       NA |       NA |     ? |       ? |      - |     - |     - |         - |
| BodyExtractorMoreComplexBody | Job-JGUMEI | .NET Core 3.1 | netcoreapp3.1 | 15.40 us | 0.252 us | 0.210 us |     ? |       ? | 0.0381 |     - |     - |    6752 B |

```